geth/peers/peerpool_test.go

```diff
@@ -136,50 +142,48 @@ func (s *PeerPoolSimulationSuite) TestSingleTopicDiscoveryWithFailover() {
 	cache, err := newInMemoryCache()
 	s.Require().NoError(err)
 	peerPool := NewPeerPool(config, cache, peerPoolOpts)
+
+	// create and start topic registry
 	register := NewRegister(topic)
-	s.Require().NoError(register.Start(s.peers[0]))
-	// need to wait for topic to get registered, discv5 can query same node
-	// for a topic only once a minute
+	err = register.Start(s.peers[0])
+	s.Require().NoError(err)
+
+	// subscribe for peer events before starting the peer pool
 	events := make(chan *p2p.PeerEvent, 20)
 	subscription := s.peers[1].SubscribeEvents(events)
 	defer subscription.Unsubscribe()
-	s.NoError(peerPool.Start(s.peers[1]))
+
+	// start the peer pool
+	s.Require().NoError(peerPool.Start(s.peers[1]))
 	defer peerPool.Stop()
 	s.Equal(signal.EventDiscoveryStarted, s.getPoolEvent(poolEvents))
-	connected := s.getPeerFromEvent(events, p2p.PeerEventTypeAdd)
-	s.Equal(s.peers[0].Self().ID, connected)
+	connectedPeer := s.getPeerFromEvent(events, p2p.PeerEventTypeAdd)
+	s.Equal(s.peers[0].Self().ID, connectedPeer)
+	// as the upper limit was reached, Discovery should be stoped
 	s.Equal(signal.EventDiscoveryStopped, s.getPoolEvent(poolEvents))
-	s.Require().Nil(s.peers[1].DiscV5)
-
-	s.Require().Equal(signal.EventDiscoverySummary, s.getPoolEvent(poolEvents))
-	summary := <-summaries
-	s.Len(summary, 1)
+	s.Equal(signal.EventDiscoverySummary, s.getPoolEvent(poolEvents))
+	s.Len(<-summaries, 1)
 
+	// stop topic register and the connected peer
 	register.Stop()
 	s.peers[0].Stop()
-	disconnected := s.getPeerFromEvent(events, p2p.PeerEventTypeDrop)
-	s.Equal(connected, disconnected)
-
-	s.Require().Equal(signal.EventDiscoverySummary, s.getPoolEvent(poolEvents))
-	summary = <-summaries
-	s.Len(summary, 0)
-
+	disconnectedPeer := s.getPeerFromEvent(events, p2p.PeerEventTypeDrop)
+	s.Equal(connectedPeer, disconnectedPeer)
+	s.Equal(signal.EventDiscoverySummary, s.getPoolEvent(poolEvents))
+	s.Len(<-summaries, 0)
+	// Discovery should be restarted because the number of peers dropped
+	// below the lower limit.
 	s.Equal(signal.EventDiscoveryStarted, s.getPoolEvent(poolEvents))
-	s.Require().NotNil(s.peers[1].DiscV5)
-	s.Require().NoError(register.Start(s.peers[2]))
+
+	// register the second peer
+	err = register.Start(s.peers[2])
+	s.Require().NoError(err)
 	defer register.Stop()
 	s.Equal(s.peers[2].Self().ID, s.getPeerFromEvent(events, p2p.PeerEventTypeAdd))
-
+	// Discovery can be stopped again.
 	s.Equal(signal.EventDiscoveryStopped, s.getPoolEvent(poolEvents))
 	s.Require().Equal(signal.EventDiscoverySummary, s.getPoolEvent(poolEvents))
-
-	summary = <-summaries
-	s.Len(summary, 1)
-
-	// verify that we are actually using cache
-	cachedPeers := peerPool.cache.GetPeersRange(topic, 1)
```

Comment:
cache got removed

Reply:
Oh yeah, because this test does everything otherwise. I will write a new one which focuses on testing cache usage.

Reply:
New tests were added in https://github.com/status-im/status-go/pull/992/commits/8651a1d725eec0a089876dc56df82da94a5212fa.